File: Construct binary tree from inorder and postorder.py
# TC-O(n)
# SC-O(n+height)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if (len(inorder) or len(postorder) )==0:
            return []
        self.dic ={}
        for i in range(len(inorder)):
            self.dic[inorder[i] ] =i
        instart =0
        inend =len(inorder ) -1
        self.postinx =len(inorder ) -1
        return self.helper(inorder ,postorder ,instart ,inend)

    def helper(self ,inorder ,postorder ,instart ,inend):

        if instart >inend or self.postinx<0:
            return None

        root =TreeNode(postorder[self.postinx])
        self.postinx-=1
        index =self.dic[root.val]
        root.right= self.helper(inorder ,postorder ,index +1 ,inend)
        root.left =self.helper(inorder ,postorder ,instart ,index -1)
        return root

File: ConstructBinaryTreeFromInorderAndPostorder.java
public class ConstructBinaryTreeFromInorderAndPostorder {
    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     *     int val;
     *     TreeNode left;
     *     TreeNode right;
     *     TreeNode() {}
     *     TreeNode(int val) { this.val = val; }
     *     TreeNode(int val, TreeNode left, TreeNode right) {
     *         this.val = val;
     *         this.left = left;
     *         this.right = right;
     *     }
     * }
     */
    Map<Integer, Integer> inorderMap = new HashMap<>();
    // TC: O(N) where N is length of array
    // SC: O(N + H) where N is length of array and H is max height of tree
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        for (int i = 0; i < inorder.length; i++) {
            inorderMap.put(inorder[i], i);
        }
        return helper(inorder, 0, inorder.length - 1, postorder, 0, postorder.length - 1);
    }

    private TreeNode helper(int[] inorder, int inStart, int inEnd, int[] postorder, int postStart, int postEnd) {
        if (inStart > inEnd || postStart > postEnd) {
            return null;
        }

        int rootVal = postorder[postEnd];
        int rootIndex = inorderMap.get(rootVal);
        int leftTreeLen = rootIndex - inStart;

        TreeNode root = new TreeNode(rootVal);
        root.left = helper(inorder, inStart, rootIndex - 1, postorder, postStart, postStart + leftTreeLen - 1);
        root.right = helper(inorder, rootIndex + 1, inEnd, postorder, postStart + leftTreeLen, postEnd - 1);
        return root;
    }
}
